src/service/resources-impl.js

```diff
@@ -578,6 +590,70 @@ export class Resources {
     }
   }
 
+  /**
+   * Builds resources during prerender in the document order. Ignores children
+   * of a resource that does not allow prerendering.
+   * @private
+   */
+  dfsPrerender_() {
+    if (!this.dfsPrerenderTreeWalker_) {
+      this.dfsPrerenderTreeWalker_ = this.win.document.createTreeWalker(
+          this.ampdoc.getBody(), NodeFilter.SHOW_ELEMENT,
+          {acceptNode: this.dfsPrerenderTreeWalkerFilter_.bind(this)}, false);
+    }
+
+    while (this.buildAttemptsCount_ <= MAX_BUILDS_IN_PRERENDER &&
+        this.dfsPrerenderTreeWalker_.nextNode()) {
+      const node = this.dfsPrerenderTreeWalker_.currentNode;
+
+      // Node isn't upgraded, set the walker head to the last upgraded node, or
+      // to the walker root, so the next pass checks its status again.
+      if (!(node.isUpgraded && node.isUpgraded())) {
+        this.dfsPrerenderTreeWalker_.previousNode();
```

Comment:
We can avoid backwards-forwards-backwards movement by moving to the next node _after_ processing whether we want to build or wait till upgraded.